block/p2p.go

```diff
@@ -0,0 +1,117 @@
+package block
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/dymensionxyz/dymint/p2p"
+	"github.com/dymensionxyz/dymint/types"
+	"github.com/ipfs/go-cid"
+	"github.com/tendermint/tendermint/libs/pubsub"
+)
+
+// onNewGossipedBlock will take a block and apply it
+func (m *Manager) onNewGossipedBlock(event pubsub.Message) {
+	eventData, ok := event.Data().(p2p.P2PBlock)
+	if !ok {
+		m.logger.Error("onNewGossipedBlock", "err", "wrong event data received")
+	}
+	block := eventData.Block
+	commit := eventData.Commit
+	source := gossipedBlock
+
+	m.logger.Debug("Received new block via gossip.", "block height", block.Header.Height, "store height", m.State.Height(), "n cachedBlocks", len(m.blockCache))
+
+	ok = m.attemptCacheBlock(&block, &commit, source)
+	if !ok {
+		return
+	}
+
+	err := m.attemptApplyCachedBlocks()
+	if err != nil {
+		m.logger.Error("Applying cached blocks.", "err", err)
+	}
+}
+
+// onNewGossipedBlock will take a block and apply it
+func (m *Manager) onNewBlockSyncBlock(event pubsub.Message) {
+	eventData, ok := event.Data().(p2p.P2PBlock)
+	if !ok {
+		m.logger.Error("onNewBlockSyncBlock", "err", "wrong event data received")
+	}
+	block := eventData.Block
+	commit := eventData.Commit
+	source := blocksyncBlock
+
+	m.logger.Debug("Received new block via blocksync.", "block height", block.Header.Height, "store height", m.State.Height(), "n cachedBlocks", len(m.blockCache))
+
+	ok = m.attemptCacheBlock(&block, &commit, source)
+	if !ok {
+		return
+	}
+
+	err := m.attemptApplyCachedBlocks()
+	if err != nil {
+		m.logger.Error("Applying cached blocks.", "err", err)
+	}
+}
+
+func (m *Manager) gossipBlock(ctx context.Context, block types.Block, commit types.Commit) error {
+	m.logger.Info("Gossipping block", "height", block.Header.Height)
+	gossipedBlock := p2p.P2PBlock{Block: block, Commit: commit}
+	gossipedBlockBytes, err := gossipedBlock.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("marshal binary: %w: %w", err, ErrNonRecoverable)
+	}
+	if err := m.p2pClient.GossipBlock(ctx, gossipedBlockBytes); err != nil {
+		// Although this boils down to publishing on a topic, we don't want to speculate too much on what
+		// could cause that to fail, so we assume recoverable.
+		return fmt.Errorf("p2p gossip block: %w: %w", err, ErrRecoverable)
+	}
+
+	// adds the block to be used by block-sync protocol
+	err = m.addBlock(ctx, block.Header.Height, gossipedBlockBytes)
+	if err != nil {
+		m.logger.Error("adding block to p2p store: %w", err)
+	}
+
+	return nil
+}
+
+// addBlock store the blocks to be used by block-sync protocol and stores the content identifier created to advertise it in the P2P network
+func (m *Manager) addBlock(ctx context.Context, height uint64, gossipedBlockBytes []byte) error {
```

Comment:
IMO if u pass the store to the P2P client, much of this code can be moved to p2p package
as it not really core logic for the `BlockManager`

Reply:
done